### PR TITLE
Added Lazy loading and pagination of Included concepts in Concept Set

### PR DESCRIPTION
**Pull request by wivern**
closes OHDSI/WebAPI#497

### CODE REVIEW

File: js/app.js [hunk 1]
```diff
@@ -945,35 +971,24 @@ define([
...
+				const getExpressionPromise = conceptSetApi.getConceptSetExpression(conceptSetId);
+				$.when(getConceptSetPromise, getExpressionPromise).done((conceptSet, expression) => {
+					self.setConceptSet(conceptSet[0], expression[0].items);
+					self.currentView(viewToShow);
+          var resolvingPromise = self.resolveConceptSetExpression();
+          $.when(resolvingPromise)
+            .done(function () {
+            	self.resolveInclusionConceptCount(conceptSetId);
+              self.currentConceptSetMode(mode);
+              $('#conceptSetLoadDialog')
```
Comment (chrisknoll) [inline js/app.js]: Don't use element selectors.  Use the modal databinding, use an observable to set the modal sate: when the call back is done, set the observable to 'false' and that will hide the modal.

Comment (chrisknoll) [reply js/app.js]: Btw: I now realize that this is existing code and you aren't actually introducing these element selectors. However, if you can clean this up as part of this PR, that'd be ideal.

**Comment by wivern** [comment]
It seems that #conceptSetLoadDialog is obsolete since there are no DOM-element with such id in the whole project. Just removed it.

File: js/app.js [hunk 2]
```diff
@@ -646,6 +640,38 @@ define([
...
+            self.resolvingConceptSetExpression(false);
+            self.conceptSetInclusionCount(count);
+           }, () => {
+            self.resolvingConceptSetExpression(false);
+            document.location = '#/configure';
+          });
+      };
+			self.resolveIncludedConcepts = function(conceptSetId) {
+        if (self.currentConceptSet() && self.currentConceptSet().id == conceptSetId && self.conceptSetInclusionIdentifiers().length > 0) {
+          return $.Deferred().resolve();
```
Comment (johnSamilin) [inline js/app.js]: Why not Promise.resolve()?

Comment (chrisknoll) [reply js/app.js]: I think this is just a situation where we have legacy code using jquery based promises and we have new code introduced which uses native promises.  We should think about an upgarde to jquery 3 which follows the same semantics as the published standard of promises.

File: js/app.js [hunk 3]
```diff
@@ -945,35 +971,22 @@ define([
 				// don't load if it is already loaded or a new concept set
 				if (self.currentConceptSet() && self.currentConceptSet()
 					.id == conceptSetId) {
-					self.currentConceptSetMode(mode);
-					self.currentView(viewToShow);
+          self.currentView(viewToShow);
+          self.currentConceptSetMode(mode);
```
Comment (pavgra) [inline js/app.js]: fix indentation here and below, pls

File: js/modules/WebAPIProvider/VocabularyProvider.js [hunk 4]
```diff
@@ -43,7 +43,17 @@ define(function (require, exports) {
 			});
 			domainPromise.resolve(domains);
 		});
-	})
+	});
+
+  function getConceptRecordCount(identifiers) {
```
Comment (pavgra) [inline js/modules/WebAPIProvider/VocabularyProvider.js]: I believe, this file should have been moved to `services` folder. Could you confirm with @johnSamilin and move the file if it is true

Comment (chrisknoll) [reply js/modules/WebAPIProvider/VocabularyProvider.js]: I believe you'll have to merge/rebase this PR onto latest master in order to get the correct refactored code .

**Comment by pavgra** [comment]
@wivern , to be honest, loader looks a bit ugly:

image

**Comment by chrisknoll** [comment]
@wivern  there is 1 conflict that needs to be resolved:

```
		</div>
<<<<<<< conceptset_lazy_loading
		<div data-bind="css: {'col-xs-10': $component.facets().length > 0, 'col-xs-12': $component.facets().length == 0}">
			<div data-bind="css: tableClasses">
=======
		<div data-bind="css: {'faceted-data-table-wrapper': true, 'col-xs-10': $component.facets().length > 0, 'col-xs-12': $component.facets().length == 0}">
			<div class="facetedDataTable">
>>>>>>> master
```

I'm not sure which side has the right HTML: the PR branch side or master. Can you resolve this conflict and I can do a final test?

**Comment by chrisknoll** [comment]
Hi, another conflict was introduced as a result of the changes to the PR related to CEM 1.0.  Can you resolve this conflict... @anthonysena , can you confirm which set of dependencies shoudl be imported based on this conflict:
https://github.com/OHDSI/Atlas/pull/786/conflicts

**Comment by anthonysena** [comment]
Sorry to have missed this - looks like the conflict is now resolved?

**Comment by chrisknoll** [comment]
OK, once again we got conflicts (sorry) but this should be the last of it for the 2.5 relase.  Can you please mrege and resolve conflicts?

Note: i attempted to do this myself, but app.js and main.js have been completely refactored out to Model.js and something else, so I'm not sure all the code changes that need to move in/out of app.js and main.js. So I would appreciate the origional authors to resolve this.  Thank you.

File: js/components/atlas.cohort-editor.html [hunk 5]
```diff
@@ -1,21 +1,21 @@
 <div class="paddedWrapper">
-	<div placeholder="enter a cohort definition description here" class="divtext cohort-description" data-bind="contentEditable:model.currentCohortDefinition().description"></div>
+	<div placeholder="enter a cohort definition description here" class="divtext cohort-description" data-bind="contentEditable:$root.currentCohortDefinition().description"></div>
 	<div data-bind="eventListener: [
 									{ event: 'click', selector: '.conceptset_import', callback: handleConceptSetImport},
 									{ event: 'click', selector: '.conceptset_edit', callback: handleEditConceptSet}
 									]">
 		<!-- ko if:canEdit-->
-		<cohort-expression-editor params="expression: model.currentCohortDefinition().expression, widget: $component.cohortExpressionEditor"></cohort-expression-editor>
+		<cohort-expression-editor params="expression: $root.currentCohortDefinition().expression, widget: $component.cohortExpressionEditor"></cohort-expression-editor>
```
[the remark this answers is not in the document]
Comment (johnSamilin) [reply js/components/atlas.cohort-editor.html]: this was done in https://github.com/OHDSI/Atlas/pull/814/commits/966b81547223c0d7e4ecc6f695103d24751ddd3f

**Pull request closed without merging**